### PR TITLE
[wip] reference: rework Macros section

### PR DESCRIPTION
**Pull request by durka**
Rearrange and add content to the Macros section to make it more informative and useful.

Changelog:
- [x] add an example at the beginning
- [x] remove non-sequitur mention of `sep_token` and `non_special_token`
- [x] fix confusing definition of `tt`
- [x] add term "fragment specifier" since that's what designators are called in the book
- [ ] rewrite explanation of lockstep repetition
  - [x] fix incorrect example
  - [ ] remove "scare quotes"
- [ ] add discussion of scoping (`#[macro_use]`, `#[macro_export]`, `#[macro_reexport]`, `$crate`)
- [x] rebase to remove unrelated changes that snuck in somehow

### CODE REVIEW

File: src/doc/reference.md [hunk 1]
```diff
@@ -555,28 +555,45 @@ Users of `rustc` can define new syntax extensions in two ways:
...
+}
+```
 
-The macro expander looks up macro invocations by name, and tries each macro
-rule in turn. It transcribes the first successful match. Matching and
+Currently, macros can expand to expressions, statements, items, patterns, or
+types[^typemacros]. This distinction is inferred from the context in which the
+macro is invoked, and carefully written macros may work in multiple contexts.
+
+[^typemacros]: Gated by
```
Comment (Manishearth) [inline src/doc/reference.md]: do footnotes work in rustdoc?

File: src/doc/reference.md [hunk 2]
```diff
@@ -555,28 +555,45 @@ Users of `rustc` can define new syntax extensions in two ways:
...
+[the `type_macros` feature](https://github.com/rust-lang/rust/issues/27245).
+
+The macro expander looks up macro invocations by name, and tries each
+arm in turn. It transcribes the first successful match. Matching and
 transcription are closely related to each other, and we will describe them
 together.
 
-### Macro By Example
+### Macros By Example
```
Comment (Manishearth) [inline src/doc/reference.md]: I don't like this change, it's actually `(Macro By Example)s`, the MBE is a single unit.

File: src/doc/reference.md [hunk 3]
```diff
@@ -586,36 +603,36 @@ syntax named by _designator_. Valid designators are:
...
 repetition. The Kleene star operator consists of `$` and parentheses, optionally
 followed by a separator token, followed by `*` or `+`. `*` means zero or more
-repetitions, `+` means at least one repetition. The parentheses are not matched or
+repetitions; `+` means at least one repetition. The parentheses are not matched or
 transcribed. On the matcher side, a name is bound to _all_ of the names it
 matches, in a structure that mimics the structure of the repetition encountered
 on a successful match. The job of the transcriber is to sort that structure
 out.
 
-The rules for transcription of these repetitions are called "Macro By Example".
```
Comment (Manishearth) [inline src/doc/reference.md]: same

File: src/doc/reference.md [hunk 4]
```diff
@@ -4062,7 +4084,7 @@ the guarantee that these issues are never caused by safe code.
 * Breaking the [pointer aliasing
   rules](http://llvm.org/docs/LangRef.html#pointer-aliasing-rules)
   with raw pointers (a subset of the rules used by C)
-* `&mut T` and `&T` follow LLVM’s scoped [noalias] model, except if the `&T`
+* `&mut` and `&` follow LLVM’s scoped [noalias] model, except if the `&T`
```
Comment (Manishearth) [inline src/doc/reference.md]: Why undo this change?

Comment (durka) [inline src/doc/reference.md]: Yeah I don't know what happened, I think I had the tab open too long and someone else landed a commit to the reference. I'll rebase to get rid of it. I only meant to edit the Macros section.

File: src/doc/reference.md (hunk 1 above)
Comment (durka) [inline src/doc/reference.md]: The same syntax is used elsewhere in the reference.

File: src/doc/reference.md (hunk 2 above)
Comment (durka) [inline src/doc/reference.md]: Hmm... "Macro by Examples" doesn't sound good at all. Maybe the subtitle should be something else? "Rule-based Macros" or "Declarative Macros"?

Comment (Manishearth) [inline src/doc/reference.md]: No, just call it "Macro By Example". All the cases where you've substituted here it's fine to use it that way.

File: src/doc/reference.md [hunk 5]
```diff
@@ -555,17 +555,33 @@ Users of `rustc` can define new syntax extensions in two ways:
 
 ## Macros
 
-`macro_rules` allows users to define syntax extension in a declarative way.  We
-call such extensions "macros by example" or simply "macros" — to be distinguished
-from the "procedural macros" defined in [compiler plugins][plugin].
+`macro_rules` allows users to define syntax extensions in a declarative way.
+We call such extensions "`macro_rules!` macros" or simply "macros" — to be
+distinguished from the "procedural macros" defined in [compiler plugins][plugin].
```
Comment (LeoTestard) [inline src/doc/reference.md]: The term ‶macros by example″ has also been used at several places and might be worth emntionning.

File: src/doc/reference.md [hunk 6]
```diff
@@ -555,17 +555,33 @@ Users of `rustc` can define new syntax extensions in two ways:
...
 
-Currently, macros can expand to expressions, statements, items, or patterns.
+As shown below, the body of a `macro_rules` macro consists of one or more
+semicolon-separated arms, similar in appearance to a `match` statement. On
+the left side of the `=>` is a matcher describing the syntax that the macro
+accepts, optionally capturing parts of it for use in the transcriber, which
+is on the right side of the `=>`. (Choice of delimiters for the macro body,
+matchers, and transcribers is arbitrary.)
```
Comment (LeoTestard) [inline src/doc/reference.md]: You should explicit the term ‶transcriber″. :)

Comment (LeoTestard) [inline src/doc/reference.md]: Choice is arbitrary, but delimiters are mandatory. You might want to add that.

File: src/doc/reference.md [hunk 7]
```diff
@@ -555,17 +555,33 @@ Users of `rustc` can define new syntax extensions in two ways:
...
-rule in turn. It transcribes the first successful match. Matching and
+Currently, macros can expand to expressions, statements, items, patterns, or
+types[^typemacros]. This distinction is inferred from the context in which the
+macro is invoked, and carefully written macros may work in multiple contexts.
+
+[^typemacros]: Gated by
+[the `type_macros` feature](https://github.com/rust-lang/rust/issues/27245).
+
+The macro expander looks up macro invocations by name, and tries each
+arm in turn. It transcribes the first successful match. Matching and
```
Comment (LeoTestard) [inline src/doc/reference.md]: Well, it should once #33840 lands at last.

File: src/doc/reference.md [hunk 8]
```diff
@@ -586,17 +603,17 @@ syntax named by _designator_. Valid designators are:
 * `ty`: a [type](#types)
 * `ident`: an [identifier](#identifiers)
 * `path`: a [path](#paths)
-* `tt`: either side of the `=>` in macro rules
+* `tt`: a "token tree": either one [token](#token) or a delimited sequence
+   of token trees
```
Comment (LeoTestard) [inline src/doc/reference.md]: You might want to add a compareason with LISP or an AST to explain more precisely what a token tree is. But maybe it would become a bit heavy.

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to reopen with comments addressed!
